### PR TITLE
[TACHYON-1201] use two watermarks for space reserver

### PR DESCRIPTION
**Pull request by chen-tian**
https://tachyon.atlassian.net/browse/TACHYON-1201

### CODE REVIEW

File: common/src/main/resources/tachyon-default.properties [hunk 1]
```diff
@@ -113,7 +113,8 @@ tachyon.worker.tieredstore.levels=1
 tachyon.worker.tieredstore.level0.alias=MEM
 tachyon.worker.tieredstore.level0.dirs.path=/mnt/ramdisk
 tachyon.worker.tieredstore.level0.dirs.quota=${tachyon.worker.memory.size}
-tachyon.worker.tieredstore.level0.reserved.ratio=0.1
+tachyon.worker.tieredstore.level0.watermark.high.ratio=0.95
+tachyon.worker.tieredstore.level0.watermark.low.ratio=0.9
```
Comment (calvinjia) [inline common/src/main/resources/tachyon-default.properties]: Could we have the default values be 0.7 and 0.9?

File: servers/src/main/java/tachyon/worker/block/SpaceReserver.java [hunk 2]
```diff
@@ -52,15 +54,25 @@
...
     mStorageTierAssoc = new WorkerStorageTierAssoc(WorkerContext.getConf());
-    Map<String, Long> capOnTiers = blockManager.getStoreMeta().getCapacityBytesOnTiers();
+    Map<String, Long> capOnTiers = mBlockManager.getStoreMeta().getCapacityBytesOnTiers();
     long lastTierReservedBytes = 0;
     for (int ordinal = 0; ordinal < mStorageTierAssoc.size(); ordinal ++) {
-      String tierReservedSpaceProp =
-          String.format(Constants.WORKER_TIERED_STORE_LEVEL_RESERVED_RATIO_FORMAT, ordinal);
       String tierAlias = mStorageTierAssoc.getAlias(ordinal);
-      long reservedSpaceBytes =
+      // To-do: make sure highWatermark ratio is larger than lowWatermark ratio
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/SpaceReserver.java]: `TODO(tianc)`

File: servers/src/main/java/tachyon/worker/block/SpaceReserver.java [hunk 3]
```diff
@@ -52,15 +54,25 @@
...
+              highWatermarkRatio));
+      mHighWaterMarkInBytesOnTiers.put(tierAlias, highWatermarkInBytes);
+
+      // LowWatemark defines when to stop the space reserving process if started
+      String lowWatermarkRatio =
+          String.format(Constants.WORKER_TIERED_STORE_LEVEL_LOW_WATERMARK_RATIO_FORMAT, ordinal);
+      long reservedSpaceBytes =
+          (long) (capOnTiers.get(tierAlias)
+              * (1 - WorkerContext.getConf().getDouble(lowWatermarkRatio)));
       mBytesToReserveOnTiers.put(tierAlias, reservedSpaceBytes + lastTierReservedBytes);
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/SpaceReserver.java]: Maybe change this to match `mHigherWaterMarkInBytesOnTiers`? Or is it used in other cases as well?

File: servers/src/test/java/tachyon/worker/block/SpaceReserverTest.java [hunk 4]
```diff
@@ -68,12 +68,19 @@ public void before() throws Exception {
...
+    String lowWatermarkRatio =
+        String.format(Constants.WORKER_TIERED_STORE_LEVEL_LOW_WATERMARK_RATIO_FORMAT, 0);
+    WorkerContext.getConf().set(lowWatermarkRatio, "0.8");
+
+    highWatermarkRatio =
+        String.format(Constants.WORKER_TIERED_STORE_LEVEL_HIGH_WATERMARK_RATIO_FORMAT, 1);
+    WorkerContext.getConf().set(highWatermarkRatio, "0.9");
+    lowWatermarkRatio =
+       String.format(Constants.WORKER_TIERED_STORE_LEVEL_LOW_WATERMARK_RATIO_FORMAT, 1);
+    WorkerContext.getConf().set(lowWatermarkRatio, "0.7");
```
Comment (calvinjia) [inline servers/src/test/java/tachyon/worker/block/SpaceReserverTest.java]: Could we add some tests for the new logic?

**Comment by calvinjia** [comment]
@chen-tian Thanks for the patch, I left a few minor comments.

File: common/src/main/java/tachyon/Constants.java [hunk 5]
```diff
@@ -274,10 +274,16 @@
       "tachyon.worker.tieredstore.level%d.dirs.quota";
   /**
    * This format is used as a template to generate the property name for a given level. e.g.,
-   * {@code String.format(Constants.WORKER_TIERED_STORE_LEVEL_RESERVED_RATIO_FORMAT, level)}
+   * {@code String.format(Constants.WORKER_TIERED_STORE_LEVEL_HIGH_WATERMARK_RATIO_FORMAT, level)}
    */
-  public static final String WORKER_TIERED_STORE_LEVEL_RESERVED_RATIO_FORMAT =
-      "tachyon.worker.tieredstore.level%d.reserved.ratio";
+  public static final String  WORKER_TIERED_STORE_LEVEL_HIGH_WATERMARK_RATIO_FORMAT =
+      "tachyon.worker.tieredstore.level%d.watermark.high.ratio";
```
Comment (apc999) [inline common/src/main/java/tachyon/Constants.java]: NIT: I don't think we need `.ratio` in the end. probably just call this (and the below one)
`tachyon.worker.tieredstore.level%d.watermark.high` and `tachyon.worker.tieredstore.level%d.watermark.low`?

Comment (apc999) [inline common/src/main/java/tachyon/Constants.java]: Also, could you update `docs/Configuration-Settings.md` and `docs/Tiered-Storage-on-Tachyon.md` where the original property name was referenced.

File: servers/src/test/java/tachyon/worker/block/SpaceReserverTest.java (hunk 4 above)
Comment (apc999) [inline servers/src/test/java/tachyon/worker/block/SpaceReserverTest.java]: +1

**Comment by yupeng9** [comment]
@chen-tian , any updates on this PR?

**Comment by chen-tian** [comment]
Have been talking with Calvin. Will work on it after my demo work done for
my company. Hopefully next week. I will fix it before 0.9 is out.

**Comment by yupeng9** [comment]
SGTM. Thanks for the update!

**Pull request closed without merging**